charmhelpers/contrib/openstack/audits/__init__.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2019 Canonical Limited.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""OpenStack Security Audit code"""
+
+import collections
+from enum import Enum
+from charmhelpers.fetch.ubuntu import get_upstream_version
+import apt_pkg
+#  noqa below is because we import apt for a side-effect
+#  initializing apt_pkg.version_compare's _system
+import apt # noqa: F401,E261
+
+
+class AuditType(Enum):
+    OpenStackSecurityGuide = 1
+
+
+_audits = {}
+
+Audit = collections.namedtuple('Audit', 'func filters')
+Config = collections.namedtuple('Config', 'audit_type config_path config_file owner group permissions')
+
+
+def audit(*args):
+    """Decorator to register an audit.
+
+    These are used to generate audits that can be run on a
+    deployed system that matches the given configuration
+
+    :param args: List of functions to filter tests against
+    :type args: List[Callable(Config)]
+    """
+    def wrapper(f):
+        test_name = f.__name__
+        if _audits.get(test_name):
+            raise RuntimeError(
+                "Test name '{}' used more than once ?"
+                .format(test_name))
+        _audits[test_name] = Audit(func=f, filters=args)
+        return f
+    return wrapper
+
+
+def is_audit_type(*args):
+    """This audit is included in the specified kinds of audits."""
+    def should_run(config):
+        if config.audit_type in args:
+            return True
+        else:
+            return False
+    return should_run
+
+
+def since_package(pkg, pkg_version):
+    """This audit should be run after the specified package version (incl)."""
+    def should_run(config):
+        version = get_upstream_version(pkg)
+        if version:
+            if apt_pkg.version_compare(version, pkg_version) >= 0:
+                return True
+        return False
```

Comment:
This bit of code can be written as a shortcut:

```python
return version and (apt_pkg.version_compare(version, pkg_version) >=0)
```

However, that might be less clear, particularly if someone is not clear that `and` shortcuts and doesn't evaluate the RHS if version is 'False-y'.  Still will return a boolean.